hugegraph-server/hugegraph-dist/src/main/java/org/apache/hugegraph/dist/MemoryMonitor.java

```diff
@@ -0,0 +1,119 @@
+ * limitations under the License.
+ */
+
+package org.apache.hugegraph.dist;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryUsage;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.hugegraph.config.HugeConfig;
+import org.apache.hugegraph.config.ServerOptions;
+import org.apache.hugegraph.util.Log;
+import org.slf4j.Logger;
+
+import com.sun.management.ThreadMXBean;
+
+public class MemoryMonitor {
+
+    private static final Logger LOG = Log.logger(MemoryMonitor.class);
+    private final double MEMORY_MONITOR_THRESHOLD;
+    private final int MEMORY_MONITOR_PERIOD;
+    private final ScheduledExecutorService scheduler;
+
+    public MemoryMonitor(String restServerConf) {
+        HugeConfig restServerConfig = new HugeConfig(restServerConf);
+        MEMORY_MONITOR_THRESHOLD =
+                restServerConfig.get(ServerOptions.JVM_MEMORY_MONITOR_THRESHOLD);
+        MEMORY_MONITOR_PERIOD =
+                restServerConfig.get(ServerOptions.JVM_MEMORY_MONITOR_PERIOD);
+        scheduler = Executors.newScheduledThreadPool(1);
+    }
+
+    private void manageMemory() {
+        double memoryUsagePercentage = getMemoryUsagePercentage();
+
+        if (memoryUsagePercentage > MEMORY_MONITOR_THRESHOLD) {
+            LOG.info("JVM memory usage is '{}', exceeding the threshold of '{}'.",
+                     memoryUsagePercentage, MEMORY_MONITOR_THRESHOLD);
+            System.gc();
+            LOG.info("Trigger System.gc()");
+
+            double doubleCheckUsage = getMemoryUsagePercentage();
+            if (doubleCheckUsage > MEMORY_MONITOR_THRESHOLD) {
+                Thread targetThread = getHighestMemoryThread();
+                if (targetThread != null) {
+                    LOG.info("JVM memory usage is '{}', exceeding the threshold of '{}'.",
+                             doubleCheckUsage, MEMORY_MONITOR_THRESHOLD);
+                    targetThread.interrupt();
+                    LOG.info("Send interrupt to '{}' thread",
+                             targetThread.getName());
+                }
+            }
+        }
+    }
+
+    private double getMemoryUsagePercentage() {
+        MemoryUsage heapMemoryUsage =
+                ManagementFactory.getMemoryMXBean().getHeapMemoryUsage();
+        return (double) heapMemoryUsage.getUsed() / heapMemoryUsage.getMax();
+    }
+
+    private Thread getHighestMemoryThread() {
+        long highestMemory = 0;
+        Thread highestThread = null;
+
+        ThreadMXBean threadMXBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
+
+        for (Thread thread : Thread.getAllStackTraces().keySet()) {
+            if (thread.getState() != Thread.State.RUNNABLE ||
+                thread.getName() == null ||
+                !thread.getName().startsWith("grizzly-http-server-")) {
```

Comment:
threads whose name matches `oltp-%d-worker-%d` may also be controlled? Since they are sub threads derived from main `grizzly-http-server-` thread and they also belongs to query section.

Reply:
Apart from Kout API, there are many queries that do not produce oltp related threads